nav2_bt_navigator/CMakeLists.txt

```diff
@@ -22,16 +22,23 @@ include_directories(
   include
 )
 
-set(executable_name bt_navigator)
+set(bt_navigator_library_name bt_navigator)
+set(navigate_to_pose_executable_name bt_navigator)
+set(follow_waypoints_executable_name follow_waypoints_bt_navigator)
```

Comment:
I don't understand why this is in this package, what is this for? If you're making a higher level application to do waypoint based following where the BT navigator is a node used in a larger behavior tree, this should be in another package to properly separate concerns.

Reply:
> I don't understand why this is in this package, what is this for?

FollowWaypointsBtNavigator reads BT and Execute the FollowWaypoints action in the same way BtNavigator currently executes the NavigateToPose action.

> If you're making a higher level application

No, it's not application.  It only provides a framework for executing FollowWaypoints actions by BT.  (In addition, The use of NavigateToPose action node in BT is also user-dependent.)

Reply:
> Should this potentially be in a new package

I think that both FollowWaypointsBtNavigator and NavigateToPoseBtNavigator does as well. What do you think?

Reply:
What do you mean both provided as plugins? For your follow case, you still need the normal BT navigator running under the hood for the `NavigateToPose` action server to follow the individual waypoints. Its not an either-or situation, this builds on top of that. 

"navigate" is how you get from A to B. A "navigator" is the thing responsible for getting you from A to B. If you're going from A to B to C to ..., that's not a navigator, that's a simple form of autonomy, where its autonomously navigating a circuit made up of individual navigation requests.

Reply:
> For your follow case, you still need the normal BT navigator running under the hood for the NavigateToPose action server to follow the individual waypoints.

Whether or not FollowWaypointsBtNavigator uses NavigateToPose is dependent on BT.

Reply:
> "navigate" is how you get from A to B. A "navigator" is the thing responsible for getting you from A to B. If you're going from A to B to C to ..., that's not a navigator, that's a simple form of autonomy, where its autonomously navigating a circuit made up of individual navigation requests.

Thanks. If "navigate" is defined that way, then the structure and naming of this PR is wrong.

Reply:
> What do you mean both provided as plugins?

I was considered to make the following structure.
- nav2_bt_navigator/bt_navigator.cpp/.hpp: Load a BT and provide interface to provide an action

plugins for bt_navigator
- nav2_bt_navigator/plugins/navigate_to_pose.cpp/.hpp: Load a BT to execute NavigateToPose  and provide an NavigateToPose 
- nav2_bt_navigator/plugins/follow_waypoints.cpp/.hpp: Load a BT to execute FollowWaypoints and provide an FollowWaypoints

Reply:
> If you're going from A to B to C to ..., that's not a navigator, that's a simple form of autonomy, where its autonomously navigating a circuit made up of individual navigation requests.

- "Moving from A to B" is "navigate"
- "Moving from A to B while avoiding obstacles" is "navigate"
- "Moving from A to B via C" is "not navigate" ?
- "Moving from A to B via C (if there is an obstacle in C, move directly to B)" is "not navigate" ?

Reply:
A to B via C could be "navigate", yes, if its a via point or a constraint on the navigation request that you go through that point. However, I would not say that going to 3 distinct points A-B-C is "navigate" if they're distinct points for distinct reasons. If B though is a point needing to go through, without stopping, as a constraint on the request for some reason, then that would be 1 navigation request. This is all semantics though. 

What I'm trying to communicate is that I don't believe this should all be in 1 package and the naming is confusing. What you're doing is creating an application level behavior tree to implement a waypoint following task. There's absolutely no argument on my side that it is a useful demo application and some folks like yourself may choose to use it as their basic autonomy. My argument is how its being added to the project (or if it belongs in this project at all). 

My assertion is that this is a demo-autonomy feature similar to the `nav2_waypoint_follower`. It uses the behavior tree rather than an action server to call the individual navigation requests that make up its parts. I think it should be either in the nav2_waypoint_follower package, since it itself is a form of waypoint following, or as another new package doing `nav2_bt_waypoint_follower` or something. But I also do wonder if this belongs better in the `nav2_tutorials` or `nav2_demos` repos.

Reply:
@fmrico do you agree with my assessment here? You also make application-level behavior trees for your research. Does this fall in line with that or am I not seeing this in the right lens?

Reply:
I almost agree with you. I'll try to create `nav2_bt_waypoint_follower`.

Reply:
Done.

Reply:
@SteveMacenski, I agree with you. This PR should be at the application-level.

@ymd-stella is contributing to #1651 with an alternative to `nav2_waypoint_follower`, but using a BT. That's great. But it shouldn't affect core packages, like `bt_navigator.` This PR could be in tutorials or demos, in a separate package. If this demo requires changes in `bt_navigator`, maybe you could do it in different and concrete PRs, in which we could discuss its convenience, but not linked to a specific demo.

In a glance, I have many doubts in the XML files that @ymd-stella proposes. Many nodes are not implemented yet. I also consider that BT shouldn't contain nodes that do what it should be done in code. They must be of a higher level and simpler.

Reply:
> If this demo requires changes in bt_navigator, maybe you could do it in different and concrete PRs, in which we could discuss its convenience, but not linked to a specific demo.

I'm thinking of changing nav2_bt_waypoint_follower to avoid depending on nav2_bt_navigator. If duplicate code is a problem for us, we will consider addressing it afterwards.

Reply:
I' ve removed the dependency from nav2_bt_waypoint_follower to nav2_bt_navigator.

Reply:
> Many nodes are not implemented yet.

Only two nodes.
- AddInt
- SetCurrentGoal

Reply:
I don't really mind how many nodes are proposed but not implemented, you're not going to be able to merge things into **production** that don't work.

Reply:
You need to provide unit and integration testing for this.

Reply:
Of course I don't think I can merge this yet. I was just pointing this out because I felt that the implemented nodes were recognized differently.

Reply:
> You need to provide unit and integration testing for this.

I'll try it.

Reply:
Resolved this conversation?

Reply:
Got it, I was adding it here since we were having a thematic conversation here, we can move to the main thread now